src/pages/explore/BiddersPage/BidVouch.tsx

```diff
@@ -1,71 +1,86 @@
+import BN from 'bn.js'
 import { useState, useEffect } from 'react'
 import { Spinner, Tab, Nav, Form, Button, InputGroup, FormControl } from 'react-bootstrap'
 import styled from 'styled-components'
 import { CurrentRound } from '../../../components/rotation-bar/CurrentRound'
 import { useKusama } from '../../../kusama'
+import { bid, vouch } from './helper'
 
-type BidVouchProps = { handleResult: any, activeAccount: accountType, accounts: accountType[] }
+type BidVouchProps = { handleResult: any, activeAccount: accountType }
+type OnStatusChangeProps = { loading: boolean, message: string, success: boolean }
 
-const BidVouch = ({ handleResult, activeAccount } : BidVouchProps) => {
+const ksmMultiplier = new BN(1e12)
+
+const BidVouch = ({ handleResult, activeAccount }: BidVouchProps) => {
   const { api, apiState } = useKusama()
-  const [bidAmount, setbidAmount] = useState(0)
+  const [bidAmount, setbidAmount] = useState<BN>(new BN(-1))
+  const [vouchValue, setVouchValue] = useState<BN>(new BN(-1))
+  const [vouchTip, setVouchTip] = useState<BN>(new BN(-1))
+  const [vouchAddress, setVouchAddress] = useState<string>()
   const [loading, setLoading] = useState(false)
 
   const apiReady = apiState === 'READY'
 
+  const onStatusChange = ({ loading, message, success }: OnStatusChangeProps) => {
+    setLoading(loading)
+    handleResult({ message, success })
+  }
+
   useEffect(() => {
-    const bid = async () => {
-      const bid = api?.tx?.society?.bid(bidAmount)
-      const injector = await web3FromAddress(activeAccount.address)
-
-      bid?.signAndSend(activeAccount.address, { signer: injector.signer }, ({ status }) => {
-        if (status.isFinalized) {
-          handleResult({ message: 'Bid submitted successfully. You are now a Bidder' , success: true })
-          setLoading(false)
-        } else {
-          handleResult({ message: `Bid request sent. Waiting for response...`, success: true })
-          setLoading(true)
-        }
-
-      }).catch((err) => {
-        setLoading(false)
-        handleResult({ message: err.message, success: false })
-      })
+    if (apiReady && bidAmount.toNumber() >= 0) {
+      const tx = api?.tx?.society?.bid(bidAmount)
+      bid(tx, activeAccount, onStatusChange)
     }
-
-    if (bidAmount > 0 && apiReady) bid()
   }, [bidAmount, handleResult])
 
-  const handleBidSubmit = async (e : React.FormEvent<HTMLFormElement>) => {
-    const bidVal = parseFloat((e.currentTarget[0] as HTMLInputElement).value)
-    setbidAmount(bidVal)
+  useEffect(() => {
+    if (apiReady && vouchAddress && vouchTip.toNumber() >= 0 && vouchValue.toNumber() >= 0) {
+      const tx = api?.tx?.society?.vouch(vouchAddress, vouchValue, vouchTip)
+      vouch(tx, activeAccount, onStatusChange)
+    }
+  }, [vouchAddress, vouchTip, vouchValue])
+
+  const handleBidSubmit = (e: React.FormEvent<HTMLFormElement>) => {
+    const bidVal : BN = new BN((e.currentTarget[0] as HTMLInputElement).value)
+    // @ts-ignore
```

Comment:
Typescript was trolling me in a way that I just gave up. It does not recognize BN as a valid type for the + operation to be applied, so I had to ignore this error since it obviously is.

Reply:
It's weird that in fixture code it worked (which I don't know why since neither JS nor TS have operator overloading). Anyway, to make sure we have no surprises in the future, I replaced all operator usages with `mul`, including in fixture code.